File: CodingChallenges/CC_02_MengerSponge_py/Box.py
# Daniel Shiffman - Translated  to Python by Ali Al Saqban
# http://codingtra.in
# http://patreon.com/codingtrain

# Menger Sponge Fractal
# https://youtu.be/LG8ZK-rRkXo

class Box(object):
    def __init__(self,x,y,z,r):
        self.pos = PVector(x,y,z)
        self.r = r
    
    def generate(self):
        
        newR = self.r/3
        
        boxes = [ Box(self.pos.x+x*newR, self.pos.y+y*newR, self.pos.z+z*newR,newR) 
                  for x in range(-1,2)
                  for y in range(-1,2)
                  for z in range(-1,2)
                  if sum(map(abs,[x,y,z])) > 1 ]

        return boxes
    
    def show(self):
        pushMatrix()
        
        translate(self.pos.x,self.pos.y,self.pos.z)
        noStroke()
        fill(255)
        box(self.r)
        
        popMatrix()

File: CodingChallenges/CC_002_MengerSponge/P5/box.js
// Daniel Shiffman
// http://codingtra.in
// http://patreon.com/codingtrain
// Code for this video: https://youtu.be/LG8ZK-rRkXo

function Box(x, y, z, r) {
  this.pos = createVector(x, y, z);
  this.r = r;

  this.generate = function() {
    let boxes = [];
    for (let x = -1; x < 2; x++) {
      for (let y = -1; y < 2; y++) {
        for (let z = -1; z < 2; z++) {
          let sum = abs(x) + abs(y) + abs(z);
          let newR = this.r / 3;
          if (sum > 1) {
            let b = new Box(
              this.pos.x + x * newR,
              this.pos.y + y * newR,
              this.pos.z + z * newR,
              newR
            );
            boxes.push(b);
          }
        }
      }
    }
    return boxes;
  };

  this.show = function() {
    push();
    translate(this.pos.x, this.pos.y, this.pos.z);
    box(this.r);
    pop();
  };
}
